File: asg1/IntList.java
// Figure A.11

public class IntList{
	protected int element;
	protected IntList next;
	public static final IntList	nil = null;
	public static int first(IntList aList){
		return aList.element;
	}
	public static IntList rest(IntList aList){
		return aList.next;
	}
	public static IntList cons(int newElement, IntList oldList){
		return new IntList(newElement, oldList);
	}
	// the real constructor, but we want cons() for interface.
	protected IntList(int newElement, IntList oldList){
		element = newElement;
		next = oldList;
	}

	/** Convert IntList to String, similar to prolog, ML style. */
	public String toString(){
		return "[" + toStringR("", this);
	}

	static String toStringR(String prefix, IntList L){
		String s;

		if (L == nil){
			s = "]";
		}else{
			s = prefix + first(L) + toStringR(", ", rest(L));
		}
		return s;
	}
}

File: asg1/intList.c
//===============================================//
//  Name:        Aidan Gadberry					 //
//  User ID:     agadberr						 //
//  Date:        4/15/2015						 //
//  Program:     Programming Assignment 1		 //
//===============================================//

#include <stdlib.h>
#include "intList.h"

// ---------- Structs -------------------------- //

// the struct of the IntList, making both the element at the node
// and the location of the next node accessible
struct IntListNode{
	int element;
	IntList next;
};

// defining constant intNil to be NULL
const IntList intNil = NULL;

// ---------- Constructors ---------------------- //

// constructs a new IntList node with element value newElement that
// connects to the front of the oldList
// Pre: newElement is int
// Post: return the new IntList at the head of the linked list, with
//       values newList->element = newElement and newList->next =
//       the oldList, or intNil if intCons was called on an intNil
IntList intCons(int newElement, IntList oldList){
	IntList newList = calloc(1, sizeof(struct IntListNode));
	newList->element = newElement;
	newList->next = oldList;
	return newList;
}

// ---------- Access Functions ------------------ //

// returns the element in the given IntList
// Pre: aList != NULL
// Post: aList still exists, aList->element is returned but remains
//       unchanged, aList->next unchanged
int intFirst(IntList aList){
	return aList->element;
}

// ---------- Manipulation Procedures ----------- //

// essentially pops off the first IntList by returning aList->next
// Pre: aList != NULL
// Post: original aList is removed from linked list, return pointer
//       to aList with unchanged aList->element and aList->next
IntList intRest(IntList aList){
	return aList->next;
}